typescript/lancedb-rag/index.ts

```diff
@@ -0,0 +1,196 @@
+import "dotenv/config";
+
+import fs from "node:fs";
+import readline from "node:readline/promises";
+import { OpenAIEmbeddingFunction, connect, type Connection } from "vectordb";
+import OpenAI from "openai";
+
+import { LiteralClient } from "@literalai/client";
+
+const DATA_URL = `https://datasets-server.huggingface.co/rows?dataset=jamescalam%2Fyoutube-transcriptions&config=default&split=train&offset=0&length=100`;
+
+const RAW_DATA_PATH = "data/raw-data.json";
+
+type RawRow = {
+  id: string;
+  title: string;
+  published: string;
+  url: string;
+  video_id: string;
+  channel_id: string;
+  text: string;
+  start: number;
+  end: number;
+};
+
+/**
+ * Get raw data
+ * Download them if from the Hugging Face Datasets server
+ * Then store locally for latter use.
+ */
+const getRawData = async (): Promise<{ rows: { row: RawRow }[] }> => {
+  if (fs.existsSync(RAW_DATA_PATH)) {
+    const rawData = fs.readFileSync(RAW_DATA_PATH, "utf8");
+    return JSON.parse(rawData);
+  }
+  const response = await fetch(DATA_URL);
+  const content = await response.text();
+  fs.writeFileSync(RAW_DATA_PATH, content, "utf8");
+  return JSON.parse(content);
+};
```

Comment:
:peanuts: point of personal preference

```suggestion
const getRawData = async (): Promise<{ rows: { row: RawRow }[] }> => {
  if (!fs.existsSync(RAW_DATA_PATH)) {
    const response = await fetch(DATA_URL);
    const content = await response.text();
    fs.writeFileSync(RAW_DATA_PATH, content, "utf8");
  }

  const rawData = fs.readFileSync(RAW_DATA_PATH, "utf8");
  return JSON.parse(rawData);
};
```

Reply:
This is one more `fs` call, because when you fetch the data, you shouldn't have to read them again.

Reply:
it's pretty inconsequential in this case but yeah i get it